File: app/src/main/java/net/nashihara/naroureader/presenter/SearchRecyclerPresenter.java
package net.nashihara.naroureader.presenter;

import android.support.v4.util.Pair;
import android.util.Log;

import com.google.firebase.crash.FirebaseCrash;

import net.nashihara.naroureader.entities.NovelItem;
import net.nashihara.naroureader.entities.Query;
import net.nashihara.naroureader.views.SearchRecyclerView;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import narou4j.Narou;
import narou4j.entities.Novel;
import narou4j.entities.NovelRank;
import narou4j.enums.NovelGenre;
import narou4j.enums.NovelType;
import narou4j.enums.OutputOrder;
import narou4j.enums.SearchWordTarget;
import rx.Emitter;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

import static narou4j.enums.OutputOrder.BOOKMARK_COUNT;
import static narou4j.enums.OutputOrder.CHARACTER_LENGTH_ASC;
import static narou4j.enums.OutputOrder.CHARACTER_LENGTH_DESC;
import static narou4j.enums.OutputOrder.HYOKA_COUNT;
import static narou4j.enums.OutputOrder.HYOKA_COUNT_ASC;
import static narou4j.enums.OutputOrder.IMPRESSION_COUNT;
import static narou4j.enums.OutputOrder.NCODE_DESC;
import static narou4j.enums.OutputOrder.OLD;
import static narou4j.enums.OutputOrder.REVIEW_COUNT;
import static narou4j.enums.OutputOrder.TOTAL_POINT;
import static narou4j.enums.OutputOrder.TOTAL_POINT_ASC;
import static narou4j.enums.OutputOrder.WEEKLY_UU;

public class SearchRecyclerPresenter implements Presenter<SearchRecyclerView> {

    private final static String TAG = SearchRecyclerPresenter.class.getSimpleName();

    private SearchRecyclerView view;

    public SearchRecyclerPresenter(SearchRecyclerView view) {
        attach(view);
    }

    @Override
    public void attach(SearchRecyclerView view) {
        this.view = view;
    }

    @Override
    public void detach() {
        view = null;
    }

    private void error(Throwable throwable) {
        Log.e(TAG, "error: ", throwable);
        FirebaseCrash.report(throwable);
        view.showError();
    }

    public void searchNovel(Query query, ArrayList<Integer> genreList) {
        buildSearchObservable(query, genreList)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(novels -> {
                if (novels == null) {
                    Log.d(TAG, "onNext: novels are null");
                    view.showRecyclerView(null);
                    return;
                }

                if (novels.size() == 0) {
                    Log.d(TAG, "onNext: novels size is zero");
                    view.showRecyclerView(null);
                    return;
                }

                List<NovelItem> novelItemList = setupNovelItems(novels);
                view.showRecyclerView(novelItemList);
            }, this::error);
    }

    private List<NovelItem> setupNovelItems(List<Novel> novelList) {
        List<NovelItem> novelItemList = new ArrayList<>();
        for (int i = 0; i < novelList.size(); i++) {
            Novel novel = novelList.get(i);
            NovelItem item = new NovelItem();
            NovelRank rank = new NovelRank();

            novel.setNcode(novel.getNcode().toLowerCase());

            rank.setNcode(novel.getNcode());
            rank.setRank(i);

            item.setNovelDetail(novel);
            item.setRank(rank);

            novelItemList.add(item);
        }

        return novelItemList;
    }

    private Observable<List<Novel>> buildSearchObservable(Query query, ArrayList<Integer> genreList) {

        return Observable.fromEmitter(emitter -> {
            Narou narou = new Narou();

            emitSearchObservable(emitter, narou, query, genreList);

            emitter.onCompleted();
        }, Emitter.BackpressureMode.NONE);
    }

    private void emitSearchObservable(Emitter<List<Novel>> emitter, Narou narou, Query query, ArrayList<Integer> genreList) {
        if (query.getNcode().equals("")) {
            emitSearchObservableFromSearchQuery(emitter, narou, query, genreList);
        } else {
            emitSearchObservableFromNcode(emitter, narou, query.getNcode());
        }
    }

    private void emitSearchObservableFromNcode(Emitter<List<Novel>> emitter, Narou narou, String ncode) {
        Novel novel = null;
        try {
            novel = narou.getNovel(ncode);
        } catch (IOException e) {
            emitter.onError(e);
        }

        if (novel == null) {
            emitter.onNext(null);
        } else {
            List<Novel> list = new ArrayList<>();
            list.add(novel);
            emitter.onNext(list);
        }
    }

    private void emitSearchObservableFromSearchQuery(
      Emitter<List<Novel>> emitter, Narou narou, Query query, ArrayList<Integer> genreList) {
        setTextParam(narou, query.getSearch(), query.getNotSearch());
        setPageLimit(narou, query.getLimit());
        setSortOrder(narou, query.getSortOrder());
        setTime(narou, query.getTime());
        setSummary(narou, query.isTargetTitle(), query.isTargetStory(), query.isTargetKeyword(), query.isTargetWriter());
        setCharLength(narou, query.getMinLength(), query.getMaxLength());
        setSerializationInfo(narou, query.isEnd(), query.isStop(), query.isPickup());
        setGenre(narou, genreList);

        Pair<List<Novel>, Throwable> result = fetchNovel(narou);
        if (result.second != null) {
            emitter.onError(result.second);
        } else {
            emitter.onNext(result.first);
        }
    }

    private void setTextParam(Narou narou, String search, String notSearch) {
        if (!search.equals("")) {
            narou.setSearchWord(search);
        }

        if (!notSearch.equals("")) {
            narou.setNotWord(notSearch);
        }
    }

    private void setPageLimit(Narou narou, int limit) {
        if (limit == 0) {
            narou.setLim(50);
        } else {
            narou.setLim(limit);
        }
    }

    private void setSummary(Narou narou, boolean title, boolean story, boolean keyword, boolean writer) {
        if (title) {
            narou.setSearchTarget(SearchWordTarget.TITLE);
        }
        if (story) {
            narou.setSearchTarget(SearchWordTarget.SYNOPSIS);
        }
        if (keyword) {
            narou.setSearchTarget(SearchWordTarget.KEYWORD);
        }
        if (writer) {
            narou.setSearchTarget(SearchWordTarget.WRITER);
        }
    }

    private void setCharLength(Narou narou, int min, int max) {
        if (min != 0 || max != 0) {
            narou.setCharacterLength(min, max);
        }
    }

    private void setSerializationInfo(Narou narou, boolean end, boolean stop, boolean pickup) {
        if (end) {
            narou.setNovelType(NovelType.ALL_NOVEL);
        } else {
            narou.setNovelType(NovelType.ALL_SERIES);
        }

        if (stop) {
            narou.setExcludeStop(true);
        }

        if (pickup) {
            narou.setPickup(true);
        }
    }

    private void setGenre(Narou narou, List<Integer> genreList) {
        if (genreList.size() > 0) {
            for (Integer genre : genreList) {
                narou.setGenre(NovelGenre.valueOf(genre));
            }
        }
    }

    private Pair<List<Novel>, Throwable> fetchNovel(Narou narou) {
        List<Novel> novels = null;
        Throwable throwable = null;
        try {
            novels = narou.getNovels();
        } catch (IOException e) {
            throwable = e;
        }

        if (novels != null) {
            novels.remove(0);
        }

        return Pair.create(novels, throwable);
    }

    private void setSortOrder(Narou narou, int order) {
        int index = order - 1;
        OutputOrder[] orders = {
          BOOKMARK_COUNT,
          REVIEW_COUNT,
          TOTAL_POINT,
          TOTAL_POINT_ASC,
          IMPRESSION_COUNT,
          HYOKA_COUNT,
          HYOKA_COUNT_ASC,
          WEEKLY_UU,
          CHARACTER_LENGTH_ASC,
          CHARACTER_LENGTH_DESC,
          NCODE_DESC,
          OLD,
        };

        if (index < orders.length) {
            narou.setOrder(orders[index]);
        }
    }

    private void setTime(Narou narou, int time) {
        int index = time - 1;
        int times[][] = {
          {0, 5},
          {5, 10},
          {10, 30},
          {30, 60},
          {60, 120},
          {120, 180},
          {180, 240},
          {240, 300},
          {300, 360},
          {360, 420},
          {420, 480},
          {480, 540},
          {540, 600},
          {600, 0},
        };

        if (index < times.length) {
            narou.setReadTime(times[index][0], times[index][1]);
        }
    }
}

File: app/src/main/java/net/nashihara/naroureader/presenter/SearchRecyclerPresenter.kt
package net.nashihara.naroureader.presenter

import android.util.Log

import com.google.firebase.crash.FirebaseCrash
import kotlinx.coroutines.experimental.Job

import net.nashihara.naroureader.entities.NovelItem
import net.nashihara.naroureader.entities.Query
import net.nashihara.naroureader.views.SearchRecyclerView

import java.util.ArrayList

import narou4j.Narou
import narou4j.entities.Novel
import narou4j.entities.NovelRank
import narou4j.enums.NovelGenre
import narou4j.enums.NovelType
import narou4j.enums.OutputOrder
import narou4j.enums.SearchWordTarget
import net.nashihara.naroureader.async
import net.nashihara.naroureader.ui

class SearchRecyclerPresenter(view: SearchRecyclerView) : Presenter<SearchRecyclerView> {

    private var view: SearchRecyclerView? = null

    private val jobList = mutableListOf<Job>()

    init {
        attach(view)
    }

    override fun attach(view: SearchRecyclerView) {
        this.view = view
    }

    override fun detach() {
        view = null
        jobList.forEach { it.cancel() }
    }

    private fun error(throwable: Throwable) {
        Log.e(TAG, "error: ", throwable)
        FirebaseCrash.report(throwable)
        view?.showError()
    }

    fun searchNovel(query: Query?, genreList: ArrayList<Int>?) {
        if (query == null || genreList == null) {
            error(Throwable("query or genreList is null"))
            return
        }

        ui {
            try {
                val novels = async { fetchNovelList(query, genreList) }.await()
                val novelItemList = setupNovelItems(novels)
                view?.showRecyclerView(novelItemList)
            } catch (e: Exception) {
                error(e)
            }
        }
    }

    private fun setupNovelItems(novelList: List<Novel>): List<NovelItem> {
        val novelItemList = ArrayList<NovelItem>()
        for (i in novelList.indices) {
            val novel = novelList[i]
            val rank = NovelRank()

            novel.ncode ?: continue

            novel.ncode = novel.ncode.toLowerCase()
            rank.ncode = novel.ncode
            rank.rank = i

            novelItemList.add(NovelItem(novelDetail = novel, rank = rank))
        }

        return novelItemList
    }

    private fun fetchNovelList(query: Query, genreList: ArrayList<Int>): List<Novel> {
        val narou = Narou()
        return if (query.ncode == "") {
            fetchNovelListFromSearchQuery(narou, query, genreList)
        } else {
            fetchNovelListFromNcode(narou, query.ncode)
        }
    }

    private fun fetchNovelListFromNcode(narou: Narou, ncode: String): List<Novel> {
        val novel: Novel = narou.getNovel(ncode)
        val list = ArrayList<Novel>()
        list.add(novel)
        return list
    }

    private fun fetchNovelListFromSearchQuery(
            narou: Narou, query: Query, genreList: ArrayList<Int>): List<Novel> {
        narou.setTextParam(query.search, query.notSearch)
        narou.setPageLimit(query.limit)
        narou.setSortOrder(query.sortOrder)
        narou.setTime(query.time)
        narou.setSummary(query.isTargetTitle, query.isTargetStory, query.isTargetKeyword, query.isTargetWriter)
        narou.setCharLength(query.minLength, query.maxLength)
        narou.setSerializationInfo(query.isEnd, query.isStop, query.isPickup)
        narou.setGenre(genreList)

        return narou.novels
    }

    companion object {

        private val TAG = SearchRecyclerPresenter::class.java.simpleName
    }
}

private fun Narou.setGenre(genreList: ArrayList<Int>) {
    if (genreList.size <= 0) {
        return
    }

    for (genre in genreList) {
        this.setGenre(NovelGenre.valueOf(genre))
    }
}

private fun Narou.setSerializationInfo(end: Boolean, stop: Boolean, pickup: Boolean) {
    if (end) {
        this.setNovelType(NovelType.ALL_NOVEL)
    } else {
        this.setNovelType(NovelType.ALL_SERIES)
    }

    if (stop) {
        this.setExcludeStop(true)
    }

    if (pickup) {
        this.setPickup(true)
    }
}

private fun Narou.setCharLength(min: Int, max: Int) {
    if (min != 0 || max != 0) {
        this.setCharacterLength(min, max)
    }
}

private fun Narou.setSummary(title: Boolean, story: Boolean, keyword: Boolean, writer: Boolean) {
    if (title) {
        this.setSearchTarget(SearchWordTarget.TITLE)
    }
    if (story) {
        this.setSearchTarget(SearchWordTarget.SYNOPSIS)
    }
    if (keyword) {
        this.setSearchTarget(SearchWordTarget.KEYWORD)
    }
    if (writer) {
        this.setSearchTarget(SearchWordTarget.WRITER)
    }
}

private fun Narou.setTime(time: Int) {
    val index = time - 1
    val times = arrayOf(intArrayOf(0, 5), intArrayOf(5, 10), intArrayOf(10, 30), intArrayOf(30, 60), intArrayOf(60, 120), intArrayOf(120, 180), intArrayOf(180, 240), intArrayOf(240, 300), intArrayOf(300, 360), intArrayOf(360, 420), intArrayOf(420, 480), intArrayOf(480, 540), intArrayOf(540, 600), intArrayOf(600, 0))

    if (index >= 0 && index < times.size) {
        this.setReadTime(times[index][0], times[index][1])
    }
}

private fun Narou.setSortOrder(order: Int) {
    val index = order - 1
    val orders = OutputOrder.values()
    if (index >= 0 && index < orders.size) {
        this.setOrder(orders[index])
    }
}

private fun Narou.setPageLimit(limit: Int) {
    if (limit == 0) {
        this.setLim(50)
    } else {
        this.setLim(limit)
    }
}

private fun Narou.setTextParam(search: String, notSearch: String) {
    if (search != "") {
        this.setSearchWord(search)
    }

    if (notSearch != "") {
        this.setNotWord(notSearch)
    }
}
